File: Google/659 Split Array into Consecutive Subsequences.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetCode
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var solution = new Solution();
            solution.IsPossible(new int[] { 1, 2, 3, 5, 5, 6, 7 });
        }
    }

    public class Solution
    {
        public bool IsPossible(int[] nums)
        {
            var counter = new Dictionary<int, int>();

            foreach (var num in nums)
            {
                if (counter.ContainsKey(num))
                    counter[num]++;
                else
                    counter[num] = 1;
            }

            var ends = new Dictionary<int, int>();
            foreach (var num in nums)
            {
                if (counter[num] < 1) continue;

                counter[num]--;
                if (ends.ContainsKey(num))
                {
                    if (--ends[num] == 0) ends.Remove(num);

                    this.AddEnds(ends, num + 1);
                }
                else
                {
                    var i = 1;
                    while (i < 3)
                    {
                        var next = num + i++;
                        if (counter.ContainsKey(next) && counter[next] > 0)
                            counter[next]--;
                        else
                            return false;
                    }

                    this.AddEnds(ends, num + i);
                }
            }

            return true;
        }

        private void AddEnds(Dictionary<int, int> ends, int key)
        {
            if (ends.ContainsKey(key))
                ends[key]++;
            else
                ends.Add(key, 1);
        }
    }
}

File: Google/659 Split Array into Consecutive Subsequences.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var isPossible = function(nums) {
    let counter = {};
    for (let num of nums) {
        if (counter.hasOwnProperty(num)) {
            counter[num]++;
        } else {
            counter[num] = 1;
        }
    }

    let ends = {};
    for (let num of nums) {
        if (counter[num] === 0) continue;

        counter[num]--;
        if (ends.hasOwnProperty(num) && ends[num] > 0) {
            ends[num]--;

            let next = num + 1;
            if (ends.hasOwnProperty(next)) {
                ends[next]++;
            } else {
                ends[next] = 1;
            }
        } else {
            let incr = 1;
            while (incr < 3) {
                let next = num + incr++;
                if (counter.hasOwnProperty(next) && counter[next] > 0) {
                    counter[next]--;
                } else {
                    return false;
                }
            }

            let value = num + incr;
            if (ends.hasOwnProperty(value)) {
                ends[value]++;
            } else {
                ends[value] = 1;
            }
        }
    }

    return true;
};